rmw_fastrtps_cpp/src/publisher.cpp

```diff
@@ -0,0 +1,181 @@
+      RMW_SET_ERROR_MSG("type support not from this implementation");
+      return nullptr;
+    }
+  }
+
+  CustomPublisherInfo * info = nullptr;
+  rmw_publisher_t * rmw_publisher = nullptr;
+  eprosima::fastrtps::PublisherAttributes publisherParam;
+
+  if (!is_valid_qos(*qos_policies)) {
+    return nullptr;
+  }
+
+  // Load default XML profile.
+  Domain::getDefaultPublisherAttributes(publisherParam);
+
+  info = new (std::nothrow) CustomPublisherInfo();
+  if (!info) {
+    RMW_SET_ERROR_MSG("failed to allocate CustomPublisherInfo");
+    return nullptr;
+  }
+
+  info->typesupport_identifier_ = type_support->typesupport_identifier;
+  info->type_support_impl_ = type_support->data;
+
+  auto callbacks = static_cast<const message_type_support_callbacks_t *>(type_support->data);
+  std::string type_name = _create_type_name(callbacks);
+  if (
+    !Domain::getRegisteredType(
+      participant_info->participant, type_name.c_str(),
+      reinterpret_cast<TopicDataType **>(&info->type_support_)))
+  {
+    info->type_support_ = new (std::nothrow) MessageTypeSupport_cpp(callbacks);
+    if (!info->type_support_) {
+      RMW_SET_ERROR_MSG("Failed to allocate MessageTypeSupport");
+      goto fail;
+    }
+    _register_type(participant_info->participant, info->type_support_);
+  }
+
+  if (!participant_info->leave_middleware_default_qos) {
+    publisherParam.qos.m_publishMode.kind = eprosima::fastrtps::ASYNCHRONOUS_PUBLISH_MODE;
+    publisherParam.historyMemoryPolicy =
+      eprosima::fastrtps::rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
+  }
+
+  publisherParam.topic.topicKind =
+    keyed ? eprosima::fastrtps::rtps::WITH_KEY : eprosima::fastrtps::rtps::NO_KEY;
+  publisherParam.topic.topicDataType = type_name;
+  publisherParam.topic.topicName = _create_topic_name(qos_policies, ros_topic_prefix, topic_name);
+
+  if (!get_datawriter_qos(*qos_policies, publisherParam)) {
+    RMW_SET_ERROR_MSG("failed to get datawriter qos");
+    goto fail;
+  }
+
+  info->listener_ = nullptr;
+  if (create_publisher_listener) {
+    info->listener_ = new (std::nothrow) PubListener(info);
+    if (!info->listener_) {
+      RMW_SET_ERROR_MSG("create_publisher() could not create publisher listener");
+      goto fail;
+    }
+  }
+
+  info->publisher_ = Domain::createPublisher(
+    participant_info->participant,
+    publisherParam,
+    info->listener_);
+  if (!info->publisher_) {
+    RMW_SET_ERROR_MSG("create_publisher() could not create publisher");
+    goto fail;
+  }
+
+  info->publisher_gid = rmw_fastrtps_shared_cpp::create_rmw_gid(
+    eprosima_fastrtps_identifier, info->publisher_->getGuid());
+
+  rmw_publisher = rmw_publisher_allocate();
+  if (!rmw_publisher) {
+    RMW_SET_ERROR_MSG("failed to allocate publisher");
+    goto fail;
+  }
+  rmw_publisher->implementation_identifier = eprosima_fastrtps_identifier;
+  rmw_publisher->data = info;
+  rmw_publisher->topic_name = reinterpret_cast<char *>(rmw_allocate(strlen(topic_name) + 1));
```

Comment:
Static cast instead?

Reply:
Done in https://github.com/ros2/rmw_fastrtps/pull/312/commits/0bf81122039925b1fac3690a0d0b4494d5734de4.